### PR TITLE
Migrating optimizer to new IR (part 1)

### PR DESCRIPTION
**Pull request by gramalingam**
Migrated the core logic. 

TO DO: decide how to handle functions. Optimizer currently incorporates function-specialization. Need to choose between function-specialization and function-inlining.

### CODE REVIEW

File: onnxscript/optimizer/constant_folding_ir.py
Comment (justinchuby) [inline onnxscript/optimizer/constant_folding_ir.py:1]: Can this be private?

File: onnxscript/optimizer/constant_folding_ir.py [hunk 1]
```diff
@@ -0,0 +1,718 @@
...
+import math
+from typing import Any, Callable, Sequence, Union
+
+import numpy as np
+import onnx
+import onnx.reference.ops
+
+import onnxscript.ir as ir
+import onnxscript.ir._convenience as _convenience
+import onnxscript.ir._enums as _enums
```
Comment (justinchuby) [inline onnxscript/optimizer/constant_folding_ir.py:19]: enums are exposed in `ir`

File: onnxscript/optimizer/constant_folding_ir.py [hunk 2]
```diff
@@ -0,0 +1,718 @@
...
+import onnxscript.ir.serde as serde
```
Comment (justinchuby) [inline onnxscript/optimizer/constant_folding_ir.py:20]: Conventionally I suggest using `ir.serde` for consistency
```suggestion
```

File: onnxscript/optimizer/constant_folding_ir.py [hunk 3]
```diff
@@ -0,0 +1,718 @@
...
+logger = logging.getLogger(__name__)
+
+# "Standard" evaluators are used to perform constant-folding.
+# The API below works only for non-control-flow ops (ops without any graph-attributes).
+# This currently used ONNX's reference implementation. But we could also
+# use ORT's implementation if we want to.
+
+
+class ReferenceEvaluator:
+    def get_evaluator(self, domain: str, op: str, version: int) -> callable | None:
```
Comment (justinchuby) [inline onnxscript/optimizer/constant_folding_ir.py:55]:
```suggestion
    def get_evaluator(self, domain: str, op: str, version: int) -> Callable | None:
```

File: onnxscript/optimizer/constant_folding_ir.py [hunk 4]
```diff
@@ -0,0 +1,718 @@
...
+
+    def evaluate(self, domain: str, op: str, version: int, *args, **kwargs) -> Any:
+        logger.debug("Evaluating %s::%s", domain, op)
+        evaluator = self.get_evaluator(domain, op, version)
+        if evaluator is None:
+            return None
+        return evaluator(*args, **kwargs)
+
+
+reference_evaluator = ReferenceEvaluator()
```
Comment (justinchuby) [inline onnxscript/optimizer/constant_folding_ir.py:70]: Can this be private?

File: onnxscript/optimizer/constant_folding_ir.py [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+    if val is None:
+        return None
+    if hasattr(val, "symbolic_value"):
+        return val.symbolic_value
+    return None
+
+
+def _get_numpy_value(val: ir.Value) -> np.ndarray | None:
+    const_value = val.const_value
+    if hasattr(const_value, "numpy"):
```
Comment (justinchuby) [inline onnxscript/optimizer/constant_folding_ir.py:162]: This should always be true?

File: onnxscript/optimizer/constant_folding_ir.py [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+def _update_type(value: ir.Value, type: ir.TypeProtocol) -> None:
+    # TODO: merge types
+    value.type = type
+
+
+def _get_input_element_type(node: ir.Node, index: int) -> int:
+    input = _get_input(node, index)
+    if input is not None and input.type is not None:
+        return input.type.dtype.value
+    return _enums.DataType.UNDEFINED.value
```
Comment (justinchuby) [inline onnxscript/optimizer/constant_folding_ir.py:203]: _enums.DataType.UNDEFINED can be used as an int so `.value` may not be necessary

File: onnxscript/optimizer/constant_folding_ir.py [hunk 7]
```diff
@@ -0,0 +1,718 @@
...
+        attr = node.attributes[name]
+        if isinstance(attr, ir.AttrInt64):
+            return attr.value
+        return None
+    return default
+
+
+# TODO(rama): The following should not be necessary. Generic incremental shape-inference
+# should handle this. This essentially implements type/shape-inference for Cast op.
+@register("Cast")
```
Comment (justinchuby) [inline onnxscript/optimizer/constant_folding_ir.py:217]: Note to self: I will resume from here

File: onnxscript/optimizer/constant_folding_ir.py [hunk 8]
```diff
@@ -0,0 +1,718 @@
...
+        for old_value, new_value in zip(old_values, new_values):
+            # Propagate relevant info from old value to new value
+            # TODO(Rama): Perhaps we should merge old and new types. As of now, new
+            # values don't have type information. Note that this could be a problem
+            # for semantics-altering rewrite-rules: we should allow users to override
+            # this for such rules.
+            new_value.type = old_value.type
+            new_value.shape = old_value.shape
+            new_value.const_value = old_value.const_value
+            new_value.name = old_value.name
```
Comment (justinchuby) [inline onnxscript/optimizer/constant_folding_ir.py:639]: We can create a copy() method for nodes

File: onnxscript/optimizer/constant_folding_ir.py [hunk 9]
```diff
@@ -0,0 +1,718 @@
...
+    return None
+
+
+class ConstantFolder:
+    opset_imports: dict[str, int]
+
+    def __init__(
+        self,
+        external_data_folder: str,
+        do_shape_inference: bool,
```
Comment (justinchuby) [inline onnxscript/optimizer/constant_folding_ir.py:464]:
```suggestion
        shape_inference: bool,
```

nit: simplicity

File: onnxscript/optimizer/constant_folding_ir.py [hunk 10]
```diff
@@ -0,0 +1,718 @@
...
+            return op.Identity(result)
+    return None
+
+
+class ConstantFolder:
+    opset_imports: dict[str, int]
+
+    def __init__(
+        self,
+        external_data_folder: str,
```
Comment (justinchuby) [inline onnxscript/optimizer/constant_folding_ir.py:463]:
```suggestion
        external_data_folder: Literal["some_options", "some_other_options"],
```

Comment (justinchuby) [inline onnxscript/optimizer/constant_folding_ir.py:463]: I would support os.PathLike as well (in general for paths)

Comment (justinchuby) [reply onnxscript/optimizer/constant_folding_ir.py:463]: folder -> dir ?

File: onnxscript/optimizer/constant_folding_ir.py [hunk 11]
```diff
@@ -0,0 +1,718 @@
...
+registry: PartialEvaluatorRegistry = PartialEvaluatorRegistry()
+
+register = registry.register
+
+
+def _get_sym_value(val: ir.Value | None) -> ir.Value | None:
+    if val is None:
+        return None
+    if hasattr(val, "symbolic_value"):
+        return val.symbolic_value
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/constant_folding_ir.py:156]: The last time we discussed this, you were ok with extending IR value to support symbolic_value, or so I thought. How does updating the IR class documentation/typing annotation to add symbolic_value compare with using meta?

File: onnxscript/optimizer/constant_folding_ir.py [hunk 12]
```diff
@@ -0,0 +1,718 @@
...
+    input = _get_input(node, index)
+    if input is not None and input.type is not None:
+        return input.type.dtype.value
+    return _enums.DataType.UNDEFINED.value
+
+
+def _get_int_attribute(node: ir.Node, name: str, default: int | None = None) -> int | None:
+    if name in node.attributes:
+        attr = node.attributes[name]
+        if isinstance(attr, ir.AttrInt64):
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/constant_folding_ir.py:209]: Hmmm ... so there is no guarantee that the specialized derived classes of ir.Attr will be created? That seems potentially confusing.

File: onnxscript/optimizer/constant_folding_ir.py [hunk 13]
```diff
@@ -0,0 +1,718 @@
...
+
+
+def _get_output(node: ir.Node, index: int) -> ir.Value | None:
+    if index < len(node.outputs):
+        return node.outputs[index]
+    return None
+
+
+def _update_type(value: ir.Value, type: ir.TypeProtocol) -> None:
+    # TODO: merge types
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/constant_folding_ir.py:195]: Ideally, we want to combine the pre-existing type/shape information with newly inferred type/shape information to retain the most information (may be the exporter has more more precise shape info or may be the inference produces more precise info) ... this requires adding check for consistency between the two, raising errors, etc. also.

File: onnxscript/optimizer/constant_folding_ir.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/optimizer/constant_folding_ir.py:209]: > potentially confusing

that's true. Do you have suggestions? We can turn the specialized classes into functions that will produce Attr objects, for example.

File: onnxscript/optimizer/constant_folding_ir.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/optimizer/constant_folding_ir.py:156]: So the latest idea was: https://github.com/microsoft/onnxscript/issues/1438#issuecomment-2094560289 . 

Wondering if you have thoughts?

File: onnxscript/optimizer/constant_folding_ir.py (hunk 9 above)
Comment (gramalingam) [reply onnxscript/optimizer/constant_folding_ir.py:464]: `shape_inference` by itself sounds ambiguous to me ... it is unclear what the option means. Maybe it is just me

File: onnxscript/optimizer/constant_folding_ir.py (hunk 8 above)
Comment (gramalingam) [reply onnxscript/optimizer/constant_folding_ir.py:639]: yes, at least, we need some common utility we can call here.

File: onnxscript/optimizer/constant_folding_ir.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/optimizer/constant_folding_ir.py:156]: Making it a transient data-structure, used only by this analysis, makes sense. Will do that.

File: onnxscript/optimizer/constant_folding_ir.py (hunk 10 above)
Comment (gramalingam) [reply onnxscript/optimizer/constant_folding_ir.py]: So, this seems unused currently! Previously, it was used when converting an external tensor into a numpy array (using ONNX's numpy_helper, which only supports "str" for the external-dir incidentally). Now, this uses IR to convert to numpy array. What happens in the IR when external-tensors are converted to numpy array? Does the IR remember the external-data-folder? Or, do we need pass it as a parameter?

Comment (justinchuby) [reply onnxscript/optimizer/constant_folding_ir.py]: It is currently incomplete, we need to update the deserialize function: https://github.com/microsoft/onnxscript/issues/1701#issuecomment-2189866219

File: onnxscript/optimizer/_constant_folding.py [hunk 14]
```diff
@@ -0,0 +1,712 @@
...
+
+registry: PartialEvaluatorRegistry = PartialEvaluatorRegistry()
+
+register = registry.register
+
+
+def _get_numpy_value(val: ir.Value) -> np.ndarray | None:
+    const_value = val.const_value
+    if hasattr(const_value, "numpy"):
+        return const_value.numpy()
```
Comment (justinchuby) [inline onnxscript/optimizer/_constant_folding.py:170]: `val.const_value` should always be an `ir.TensorProto` type or None so it should always have `numpy()` when it is not None. Currently the type is not enforced for runtime efficiency, but we can add more type checking under the `DEBUG` mode

File: onnxscript/optimizer/_constant_folding.py [hunk 15]
```diff
@@ -0,0 +1,712 @@
...
+    return None
+
+
+def _get_bool_value(val: ir.Value | None) -> bool | None:
+    if val is None:
+        return None
+    val = _get_numpy_value(val)
+    if val is None:
+        return None
+    if isinstance(val, bool):
```
Comment (justinchuby) [inline onnxscript/optimizer/_constant_folding.py:180]: Curious when this is needed? Do we need to update the typing annotation or tighten the invariance?

File: onnxscript/optimizer/_constant_folding.py [hunk 16]
```diff
@@ -0,0 +1,712 @@
...
+        return node.outputs[index]
+    return None
+
+
+def _update_type(value: ir.Value, type: ir.TypeProtocol) -> None:
+    # TODO: merge types
+    value.type = type
+
+
+def _get_input_element_type(node: ir.Node, index: int) -> int:
```
Comment (justinchuby) [inline onnxscript/optimizer/_constant_folding.py:206]:
```suggestion
def _get_input_element_type(node: ir.Node, index: int) -> ir.DataType:
```

seems appropriate based on usage

File: onnxscript/optimizer/_constant_folding.py [hunk 17]
```diff
@@ -0,0 +1,712 @@
...
+    def __init__(self):
+        self.op_evaluators: dict[tuple[str, str], list[PartialEvaluator]] = {}
+
+    def lookup_evaluators(self, domain: str, opname: str, version: int):
+        evaluator_list = self.op_evaluators.get((domain, opname), [])
+        return [
+            evaluator.function for evaluator in evaluator_list if evaluator.valid_for(version)
+        ]
+
+    def register(self, opname: str, domain: str = "", version=None):
```
Comment (justinchuby) [inline onnxscript/optimizer/_constant_folding.py:140]:
```suggestion
    def register(self, opname: str, domain: str = "", version=None) -> Callable[[T], T]:
```

File: onnxscript/optimizer/_constant_folding.py [hunk 18]
```diff
@@ -0,0 +1,712 @@
...
+
+
+class OptimizerState:
+    def __init__(self):
+        self._sym_value_map: dict[ir.Value, Any] = {}
+
+    def get_sym_value(self, value: ir.Value | None) -> Any:
+        if value is None:
+            return None
+        return self._sym_value_map.get(value, None)
```
Comment (justinchuby) [inline onnxscript/optimizer/_constant_folding.py:88]:
```suggestion
        return self._sym_value_map.get(value)
```

File: onnxscript/optimizer/_constant_folding.py [hunk 19]
```diff
@@ -0,0 +1,712 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# -------------------------------------------------------------------------
```
Comment (justinchuby) [inline onnxscript/optimizer/_constant_folding.py:4]:
```suggestion
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
```

**Review by justinchuby — APPROVED** [review]
lgtm w/ comments. In particular, I think `const_value` should always be `ir.TensorProtocol` and we can rely on this invariance.

File: onnxscript/optimizer/_constant_folding.py (hunk 19 above)
Comment (gramalingam) [reply onnxscript/optimizer/_constant_folding.py]: done

File: onnxscript/optimizer/_constant_folding.py (hunk 16 above)
Comment (gramalingam) [reply onnxscript/optimizer/_constant_folding.py:206]: skipping this for now. (will require dropping the `.value` in the implementation in two places as well.)

File: onnxscript/optimizer/constant_folding_test.py [hunk 20]
```diff
@@ -3,12 +3,29 @@
...
+    def _fold(self, model: onnx.ModelProto, onnx_shape_inference=False):
+        if self.using_ir:
+            ir_model = serde.deserialize_model(model)
+            _constant_folding.fold_constants(
+                ir_model, onnx_shape_inference=onnx_shape_inference
+            )
+            optimizer.remove_unused_nodes(ir_model)
+            return serde.serialize_model(ir_model)
+        else:
+            constant_folding.fold_constants(model, onnx_shape_inference=onnx_shape_inference)
```
Comment (titaiwangms) [inline onnxscript/optimizer/constant_folding_test.py:25]: Will we replace constant_folding with _constant_folding eventually? If so, _constant_folding should not import stuffs from constant_folding?

Comment (gramalingam) [reply onnxscript/optimizer/constant_folding_test.py:25]: yes ... they will migrate into this file eventually

**Pull request merged**